DIPs/1NNN-QFS.md

```diff
@@ -0,0 +1,406 @@
+if candidates contain `enum` parameters, constant folding must be attempted for them, i.e.
+a candidate can only be excluded when an `enum` parameter is bound to an argument for which constant folding failed.
+
+A non-`enum` parameter binds to run-time and compile-time arguments,
+but for compile-time values, `enum` parameters are a better match
+(cf. lvalues and rvalues bind to non-`ref` parameters, but `ref` parameters are a better match for lvalues).
+
+An `enum` non-`static` member function template can only be called with a compile-time constant object.
+In its body, `this` semantically behaves like a template value parameter, e.g. it can be used as a template argument.
+The `enum` member function attribute is incompatible with a [template `this` parameter](https://dlang.org/spec/template.html#template_this_parameter).
+Conceptually, an `enum` non-`static` member function template behaves as if implemented as `static` member function template,
+with the only difference that the calling syntax is `obj.method(args)` instead of `typeof(obj).method!obj(args)`.
+
+In the function body (including contracts and constraints), an `enum` parameter’s value is a compile-time constant as if it were template value parameter.
+The same is true for `this` in an `enum` non-`static` member function body.
+The difference between `enum` parameters and template value parameters is only in calling syntax:
+`f!ct_value(args)` versus `f(ct_value, args)`.
+
+The `enum` storage class is incompatible with any other storage classes currently in the language except type constructors.
+Using them together is a compile error.
+This proposal does not define `enum` to be incompatible with potential future storage classes;
+it just so happens that all the current parameter storage classes have semnatics that do not apply to compile-time constants.
+<!--
+The following is wrong, but it could be because of a bug.
+Type constructors are allowed, but have no effect
+because compile-time values are essentially `immutable`,
+and for any type constructor `qual` and every type `T`, we have `qual(immutable T)` equal to `immutable T`.
+-->
+
+#### `auto enum` Storage Class
+
+One can use `auto enum` to infer from the argument wether it is a compile-time constant (cf. `auto ref` to infer the argument’s value category).
+
+An `auto enum` parameter bindy to any argument of compatible type.
+On overload resolution,
+when comparing overload candidates in the partial ordering,
+`auto enum` is a worse match than `enum` and other non-`auto enum`.
+When the best candidate determined by overload resolution contains `auto enum` parameters,
+constant folding must be attempted for them, i.e.
+`auto enum` only binds the argument as a run-time value after constant folding failed.
+
+The `auto enum` storage class is compatible with all other storage classes except `ref`,
+i.e. `auto enum auto ref` is valid, but `auto enum ref` is invalid.
```

Comment:
I think `ref` should be allowed for enum functions. You can still pass members of `enum` arrays to `ref` arguments of regular functions and assign the result to another `enum`, I just tested. If the `enum` array in question is used twice, the function won't change the contents for remaining uses, but still. While there is little point in defining an `enum` `ref` argument, it's conceivable that people might want `auto enum ref`. `enum ref` should be allowed simply because of consistency and implementation simplicity, for the same reason e.g. `nothrow int x;` is allowed.

Since `ref` should be allowed, `in`, `out`, `scope`, `return`, `return ref` and `return scope` should also be allowed.

Reply:
ping @Bolpat

Reply:
Sorry for the late answer, I didn’t have much time in the last month.
> You can still pass members of `enum` arrays to `ref` arguments of regular functions and assign the result to another `enum`, I just tested. If the `enum` array in question is used twice, the function won't change the contents for remaining uses, but still.

I honestly don’t understand what you mean. An example may help.

As to `enum ref`, function attributes (cf. `nothrow int x;` example) generally combine with any other attributes. Parameter storage classes don’t. Notably, `out` and `ref` are incompatible, `in` will be incompatible with `ref` and always was with `out`. I view `enum` in the category that `in`, `ref`, and `out` form: They define the way the parameter is passed and therefore are mutually exclusive.

Reply:
> Sorry for the late answer, I didn’t have much time in the last month.

I haven't been exactly the paragon of virtue with responding quickly either. I'm not blaming other people for that.

> I honestly don’t understand what you mean. An example may help.

```D
enum arr = [0,1,2];

@safe increment(ref int x){x++;}
@safe void main()
{   import std;
    
    arr[1].increment(); // Legal
    arr.writeln; // Still prints [0, 1, 2]
}
```

> generally combine with any other attributes. Parameter storage classes don’t.

But assuming you accept `auto enum ref` is needed, we already have a combination of `enum` and `ref`. If we don't allow `enum ref` too, it's a special case rule that feels a bit smelly to me. OTOH it's a matter of taste, there's no reason why mine would be any better than yours.